### PR TITLE
FBX-1 Fix null shared mesh

### PR DESCRIPTION
**Pull request by natalia-wh**
Fix error when exporting a game object with a missing mesh.
Reformat `ExportInstance` code and simplify instance export.

### CODE REVIEW

File: com.unity.formats.fbx/Editor/FbxExporter.cs [hunk 1]
```diff
@@ -184,7 +184,7 @@ internal enum FbxNodeRelationType
         /// <summary>
         /// Map the ID of a prefab to an FbxMesh (for preserving instances) 
         /// </summary>
-        Dictionary<int, FbxMesh> SharedMeshes = new Dictionary<int, FbxMesh> ();
+        Dictionary<Mesh, FbxNode> SharedMeshes = new Dictionary<Mesh, FbxNode>();
```
Comment (vkovec) [inline com.unity.formats.fbx/Editor/FbxExporter.cs:187]: can you please update the comment for this

File: com.unity.formats.fbx/Editor/FbxExporter.cs [hunk 2]
```diff
@@ -1389,41 +1387,32 @@ private bool ExportInstance(GameObject unityGo, FbxScene fbxScene, FbxNode fbxNo
                 return false;
             }
 
-            Object unityPrefabParent = PrefabUtility.GetCorrespondingObjectFromSource(unityGo);
```
Comment (vkovec) [inline com.unity.formats.fbx/Editor/FbxExporter.cs:1392]: interesting so the prefab parts were not necessary in the end?

File: com.unity.formats.fbx/Editor/FbxExporter.cs [hunk 3]
```diff
@@ -1389,41 +1387,32 @@ private bool ExportInstance(GameObject unityGo, FbxScene fbxScene, FbxNode fbxNo
...
+            Mesh unityGoMesh = null;
+            // get the mesh of the game object
+            if (unityGo.TryGetComponent<MeshFilter>(out MeshFilter meshFilter))
+            {
+                unityGoMesh = meshFilter.sharedMesh;
+            }
 
-            if (unityPrefabParent != null && !SharedMeshes.TryGetValue (unityPrefabParent.GetInstanceID(), out fbxMesh))
+            // export mesh as an instance if it is a duplicate mesh or a prefab
+            if (unityGoMesh && SharedMeshes.ContainsKey(unityGoMesh))
```
Comment (vkovec) [inline com.unity.formats.fbx/Editor/FbxExporter.cs:1400]: for here you can do `if(!unityGoMesh)` then return false

Comment (vkovec) [reply com.unity.formats.fbx/Editor/FbxExporter.cs:1400]: also suggestion is to use TryGetValue for SharedMeshes to get the FbxNode

File: com.unity.formats.fbx/Editor/FbxExporter.cs (hunk 2 above)
Comment (natalia-wh) [reply com.unity.formats.fbx/Editor/FbxExporter.cs:1392]: The prefab id was just used to grab an fbx mesh, which can also be done with a unity mesh. The id was never assigned to the mesh itself.

File: com.unity.formats.fbx/Editor/FbxExporter.cs [hunk 4]
```diff
@@ -1389,41 +1387,42 @@ private bool ExportInstance(GameObject unityGo, FbxScene fbxScene, FbxNode fbxNo
...
-                if (ExportMesh (unityGo, fbxNode) && fbxNode.GetMesh() != null) {
-                    SharedMeshes [unityPrefabParent.GetInstanceID()] = fbxNode.GetMesh ();
-                    return true;
-                }
                 return false;
             }
-            // check if mesh is shared between 2 objects that are not prefabs
-            else if (unityPrefabParent == null)
+            // export mesh as an instance if it is a duplicate mesh or a prefab
+            else if (unityGoMesh && SharedMeshes.ContainsKey(unityGoMesh))
```
Comment (vkovec) [inline com.unity.formats.fbx/Editor/FbxExporter.cs:1406]: if you remove the `else` should no longer need the `unityGoMesh` check here

File: com.unity.formats.fbx/Editor/FbxExporter.cs [hunk 5]
```diff
@@ -1389,41 +1387,42 @@ private bool ExportInstance(GameObject unityGo, FbxScene fbxScene, FbxNode fbxNo
...
-                    if (unityGoMesh != null)
-                    {
-                        MeshToFbxNodeMap.Add(unityGoMesh.sharedMesh, fbxNode);
-                    }
-                    return false;
+                    fbxMesh = node.GetMesh();
                 }
             }
+            // unique mesh, so save it to find future duplicates
+            else if (unityGoMesh)
```
Comment (vkovec) [inline com.unity.formats.fbx/Editor/FbxExporter.cs:1419]: same for here should be possible to remove the if condition

File: com.unity.formats.fbx/Editor/FbxExporter.cs [hunk 6]
```diff
@@ -1389,41 +1387,42 @@ private bool ExportInstance(GameObject unityGo, FbxScene fbxScene, FbxNode fbxNo
...
-                if (unityGoMesh != null && MeshToFbxNodeMap.ContainsKey(unityGoMesh.sharedMesh))
+                if (Verbose)
                 {
-                    fbxMesh = MeshToFbxNodeMap[unityGoMesh.sharedMesh].GetMesh();
+                    Debug.Log (string.Format ("exporting instance {0}", unityGo.name));
                 }
-                // export mesh as normal and add it to list
-                else
+
+                if (SharedMeshes.TryGetValue(unityGoMesh, out FbxNode node))
```
Comment (vkovec) [inline com.unity.formats.fbx/Editor/FbxExporter.cs:1413]: if you do this TryGetValue at the if statement on line 1406, you won't need to get the node here again if it is contained in the dict

File: com.unity.formats.fbx/Editor/FbxExporter.cs [hunk 7]
```diff
@@ -1389,41 +1387,40 @@ private bool ExportInstance(GameObject unityGo, FbxScene fbxScene, FbxNode fbxNo
...
+            if (!unityGoMesh)
             {
-                if (Verbose)
-                    Debug.Log (string.Format ("exporting instance {0}({1})", unityGo.name, unityPrefabParent.name));
-                
-                if (ExportMesh (unityGo, fbxNode) && fbxNode.GetMesh() != null) {
-                    SharedMeshes [unityPrefabParent.GetInstanceID()] = fbxNode.GetMesh ();
-                    return true;
-                }
+                Debug.LogWarningFormat("Warning: Could not find a valid mesh for {0}.", unityGo.name);
```
[the remark this answers is not in the document]
Comment (natalia-wh) [reply com.unity.formats.fbx/Editor/FbxExporter.cs:1403]: Ah right, will remove it then.

**Review by vkovec — APPROVED** [review]

**Pull request merged**